homeassistant/components/xiaomi_miio/fan.py

```diff
@@ -674,3 +696,184 @@ async def async_reset_filter(self):
             "Resetting the filter lifetime of the miio device failed.",
             self._device.reset_filter,
         )
+
+
+class XiaomiFan(XiaomiGenericDevice):
+    """Representation of a Xiaomi Fan."""
+
+    def __init__(self, name, device, entry, unique_id, coordinator):
+        """Initialize the plug switch."""
+        super().__init__(name, device, entry, unique_id, coordinator)
+
+        if self._model == MODEL_FAN_P5:
+            self._device_features = FEATURE_FLAGS_FAN_P5
+            self._preset_modes = [mode.name for mode in FanOperationMode]
+        else:
+            self._device_features = FEATURE_FLAGS_FAN
+            self._preset_modes = [ATTR_MODE_NATURE, ATTR_MODE_NORMAL]
+            self._nature_mode = False
+        self._supported_features = (
+            SUPPORT_SET_SPEED
+            | SUPPORT_OSCILLATE
+            | SUPPORT_PRESET_MODE
+            | SUPPORT_DIRECTION
+        )
+        self._preset_mode = None
+        self._oscillating = None
+        self._percentage = None
+
+    @property
+    def preset_mode(self):
+        """Get the active preset mode."""
+        return ATTR_MODE_NATURE if self._nature_mode else ATTR_MODE_NORMAL
+
+    @property
+    def percentage(self):
+        """Return the current speed as a percentage."""
+        return self._percentage
+
+    @property
+    def oscillating(self):
+        """Return whether or not the fan is currently oscillating."""
+        return self._oscillating
+
+    @callback
+    def _handle_coordinator_update(self):
+        """Fetch state from the device."""
+        self._available = True
+        self._state = self.coordinator.data.is_on
+        self._oscillating = self.coordinator.data.oscillate
+        self._nature_mode = self.coordinator.data.natural_speed != 0
+        if self.coordinator.data.is_on:
+            if self._nature_mode:
+                self._percentage = self.coordinator.data.natural_speed
+            else:
+                self._percentage = self.coordinator.data.direct_speed
+        else:
+            self._percentage = 0
+
+        self.async_write_ha_state()
+
+    async def async_set_preset_mode(self, preset_mode: str) -> None:
+        """Set the preset mode of the fan."""
+        if preset_mode not in self.preset_modes:
+            _LOGGER.warning("'%s'is not a valid preset mode", preset_mode)
+            return
+
+        if preset_mode == ATTR_MODE_NATURE:
+            await self._try_command(
+                "Setting fan speed percentage of the miio device failed.",
```

Comment:
```suggestion
                "Setting natural fan speed percentage of the miio device failed.",
```
Maybe change this to differentiate from the error message below?